### PR TITLE
 14-day Redis Snapshot Retention on Delete

### PR DESCRIPTION
**Pull request by rbogle**
## Changes proposed in this pull request:

-  Redis broker will on delete:
-  create a snapshot
- use the aws-broker s3 bucket with a 14day retention policy
-  copy snapshot to s3 bucket path orgid/spaceid/serviceid/instanceid
- write json file of instance info to same path
- delete redis manual backup. 

## Security considerations

None

### CODE REVIEW

File: ci/terraform/s3_module/s3.tf [hunk 1]
```diff
@@ -31,12 +36,13 @@ data "aws_iam_policy_document" "bucket_policy" {
...
 
+data "aws_canonical_user_id" "current" {}
+
 module "aws_s3_bucket"{
   source = "terraform-aws-modules/s3-bucket/aws"
   version = "2.8.0"
   bucket = local.bucket_name
-  acl    = var.s3_acl
-  
+  acl = null
```
Comment (ccostino) [inline ci/terraform/s3_module/s3.tf:45]: Is this change related to the IAM policy updates above?

File: services/redis/redis.go [hunk 2]
```diff
@@ -287,3 +294,95 @@ func (d *dedicatedRedisAdapter) didAwsCallSucceed(err error) bool {
...
+	for {
+		resp, err := ec_svc.DescribeSnapshots(check_input)
+		if success := d.didAwsCallSucceed(err); !success {
+			d.logger.Error("exportRedisSnapshot: Redis.DescribeSnapshots Failed", err, lager.Data{"uuid": i.Uuid})
+			return
+		}
+		if *(resp.Snapshots[0].SnapshotStatus) == "available" {
+			break
+		}
+		time.Sleep(sleep)
```
Comment (ccostino) [inline services/redis/redis.go:324]: Is the `sleep` call here just to make sure the operation has a chance to finish?

File: services/redis/redis.go [hunk 3]
```diff
@@ -287,3 +294,95 @@ func (d *dedicatedRedisAdapter) didAwsCallSucceed(err error) bool {
...
```
Comment (ccostino) [inline services/redis/redis.go:374]: Same question as above; just curious if this is here for allowing the operation to finish or another purpose.

**Review by ccostino — COMMENTED** [review]
Thanks, @rbogle!  Just a couple of quick questions.

Noting here for other folks that the DB changes are here because they've already been made in the environments - the pipelines were flown and the changes already made in AWS.

File: ci/terraform/s3_module/s3.tf (hunk 1 above)
Comment (rbogle) [reply ci/terraform/s3_module/s3.tf:45]: actually related to the grants down below because they are mutually exclusive. 
The grants added give elasticache in govcloud and commercial (if ever needed) access to the bucket to write snapshots.

File: services/redis/redis.go (hunk 2 above)
Comment (rbogle) [reply services/redis/redis.go:324]: yes these are polling routines that wait for the snapshot state to become' available' before moving on to the next task. 
The sleep is so that we don't get throttled/blocked by the AWS API for two many requests in a short period. 
This is all in a go routine so sleeping here doesn't block the rest of the broker.

File: services/redis/redis.go (hunk 3 above)
Comment (rbogle) [reply services/redis/redis.go:374]: yes these are polling routines that wait for the snapshot state to become' available' before moving on to the next task. 
The sleep is so that we don't get throttled/blocked by the AWS API for too many requests in a short period. 
This is all in a go routine so sleeping here doesn't block the rest of the broker.

**Review by ccostino — APPROVED** [review]
Thank you, @rbogle!
